wallet/datastore.go

```diff
@@ -431,47 +354,419 @@ func (pg *Postgres) IncreaseLinkingLimit(ctx context.Context, providerLinkingID
+	}
+	// will rollback if tx created at this scope
+	defer rollback()
+
+	// sql query to perform unlinking
+	stmt := `
+		update
+			wallets
+		set
+			user_deposit_destination=''
+		where
+			id=$1
+	`
+	// perform query
+	if r, err := tx.ExecContext(
+		ctx,
+		stmt,
+		walletID,
+	); err != nil {
+		sublogger.Error().Err(err).Msg("failed to update wallet_custodian_id for wallet")
+		return err
+	} else if r != nil {
+		count, _ := r.RowsAffected()
+		if count < 1 {
+			sublogger.Error().Msg("at least one record should be updated for disconnecting a verified wallet")
+			return errors.New("should have updated at least one wallet with disconnected custodial")
+		}
+	}
+
+	// set disconnected on the custodian link
+	stmt = `
+		update
+			wallet_custodian
+		set
+			disconnected_at=now()
+		where
+			wallet_id=$1 and disconnected_at is null
+	`
+	// perform query
+	if r, err := tx.ExecContext(
+		ctx,
+		stmt,
+		walletID,
+	); err != nil {
+		sublogger.Error().Err(err).Msg("failed to update wallet_custodian_id for wallet")
+		return err
+	} else if r != nil {
+		count, _ := r.RowsAffected()
+		if count < 1 {
+			sublogger.Error().Msg("at least one record should be updated for disconnecting a verified wallet")
+			return errors.New("should have updated at least one wallet with disconnected custodial")
+		}
+	}
+
+	// if the tx was created in this scope we will commit here
+	if err := commit(); err != nil {
+		return fmt.Errorf("failed to commit DisconnectCustodialWallet transaction: %w", err)
+	}
+
+	// done
+	return nil
+}
+
+// ConnectCustodialWallet - create a record of a custodian wallet
+func (pg *Postgres) ConnectCustodialWallet(ctx context.Context, cl *CustodianLink, depositDest string) error {
+	var err error
+	// create a sublogger
+	sublogger := logger(ctx).With().
+		Str("wallet_id", cl.GetWalletIDString()).
+		Str("custodian", cl.Custodian).
+		Str("linking_id", cl.GetLinkingIDString()).
+		Logger()
+
+	sublogger.Debug().
+		Msg("creating linking of wallet custodian")
+
+	// get tx
+	ctx, tx, rollback, commit, err := getTx(ctx, pg)
+	if err != nil {
+		return fmt.Errorf("failed to create db transaction ConnectCustodialWallet: %w", err)
+	}
+	// will rollback if tx created at this scope
+	defer rollback()
+
+	var existingLinkingID uuid.UUID
+	// get the custodial provider's linking id from db
+	stmt := `
+		select linking_id from wallet_custodian
+		where wallet_id=$1 and custodian=$2
+	`
+	err = tx.Get(&existingLinkingID, stmt, cl.WalletID, cl.Custodian)
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		sublogger.Error().Err(err).
+			Msg("failed to insert wallet_custodian")
+		return fmt.Errorf("failed to insert wallet custodian record: %w", err)
+	}
+
+	if !uuid.Equal(existingLinkingID, *new(uuid.UUID)) {
+		// check if the member matches the associated member
+		if !uuid.Equal(*cl.LinkingID, existingLinkingID) {
+			return handlers.WrapError(errors.New("wallets do not match"), "unable to match wallets", http.StatusForbidden)
```

Comment:
does the client handle this case?

Reply:
This is how we currently respond to this type of error.  I can ask if they appropriately handle, but is is outlined as a known error condition in the spec.